plugins/error-augmentor/plugin.go

```diff
@@ -0,0 +1,231 @@
+	"aspect.build/cli/pkg/ioutils"
+	"aspect.build/cli/pkg/plugin/sdk/v1alpha2/config"
+)
+
+func main() {
+	goplugin.Serve(config.NewConfigFor(NewDefaultPlugin()))
+}
+
+type ErrorAugmentorPlugin struct {
+	properties             []byte
+	hintMap                map[*regexp.Regexp]string
+	yamlUnmarshalStrict    func(in []byte, out interface{}) (err error)
+	helpfulHints           *helpfulHintSet
+	helpfulHintsMutex      sync.Mutex
+	errorMessages          chan string
+	errorMessagesWaitGroup sync.WaitGroup
+}
+
+func NewDefaultPlugin() *ErrorAugmentorPlugin {
+	return NewPlugin()
+}
+
+func NewPlugin() *ErrorAugmentorPlugin {
+	return &ErrorAugmentorPlugin{
+		properties:          nil,
+		hintMap:             map[*regexp.Regexp]string{},
+		yamlUnmarshalStrict: yaml.UnmarshalStrict,
+		helpfulHints:        &helpfulHintSet{nodes: make(map[helpfulHintNode]struct{})},
+	}
+}
+
+// will be used to unmarshal plugin properties specific to this plugin
+type errorMappings struct {
+	ErrorMappings map[string]string `yaml:"error_mappings"`
+}
+
+func (plugin *ErrorAugmentorPlugin) SetupHook(
+	properties []byte,
+) error {
+	plugin.properties = properties
+
+	var processedProperties errorMappings
+	if err := plugin.yamlUnmarshalStrict(properties, &processedProperties); err != nil {
+		return fmt.Errorf("failed to setup: failed to parse properties: %w", err)
+	}
+
+	// change map keys into regex objects now so they are ready to use and we only need to compile the regex once
+	for r, m := range processedProperties.ErrorMappings {
+		// r for regex, m for message
+		regex, err := regexp.Compile(r)
+		if err != nil {
+			return err
+		}
+
+		plugin.hintMap[regex] = m
+	}
+
+	plugin.errorMessages = make(chan string, 64)
+	for i := 0; i < 4; i++ {
+		go plugin.errorMessageProcessor()
+		plugin.errorMessagesWaitGroup.Add(1)
+	}
+
+	return nil
+}
+
+func (plugin *ErrorAugmentorPlugin) BEPEventCallback(event *buildeventstream.BuildEvent) error {
+	aborted := event.GetAborted()
+	if aborted != nil {
+		plugin.errorMessages <- aborted.Description
+
+		// We exit early here because there will not be a progress message when the event was of type "aborted".
+		return nil
+	}
+
+	progress := event.GetProgress()
+
+	if progress != nil {
+		// TODO: Should we also check for stdout?
+		stderr := progress.GetStderr()
+		if stderr != "" {
+			plugin.errorMessages <- stderr
+		}
+	}
+
+	return nil
+}
+
+func (plugin *ErrorAugmentorPlugin) errorMessageProcessor() {
+	for errorMessage := range plugin.errorMessages {
+		plugin.processErrorMessage(errorMessage)
+	}
+	plugin.errorMessagesWaitGroup.Done()
+}
+
+func (plugin *ErrorAugmentorPlugin) processErrorMessage(errorMessage string) {
+	for regex, helpfulHint := range plugin.hintMap {
+		matches := regex.FindStringSubmatch(errorMessage)
+
+		if len(matches) > 0 {
+
+			// apply regex capture group replacements to given hint
+			for i, match := range matches {
+				if i == 0 {
```

Comment:
Why are we skipping index 0 again? I don't recall. It's worth a comment?

Reply:
Oh, reading the docs from `FindStringSubmatch` refreshed my memory. I still think a comment helps clarify.